File: EnigmaMachine/Python/Plugboard.py
from string import ascii_uppercase

class plugboard:
    def __init__(self, pairs):                              # The plugboard is used to swap letters when encrypting a letter
        self.left = ascii_uppercase                         # Left is a string with the swapped letters
        self.right =  ascii_uppercase                       # While right is a refference to the alphabet

        for i in pairs:
            A = i[0]
            B = i[1]
            pos_A = self.left.find(A)
            pos_B = self.left.find(B)
            self.left = self.left[:pos_A] + B + self.left[pos_A+1:]
            self.left = self.left[:pos_B] + A + self.left[pos_B+1:]

    def forward(self, signal):                              # This method converts a position from the right the to the left strings
        letter = self.right[signal]
        signal = self.left.find(letter)
        return signal 

    def backward(self, signal):                             # This method does the exact opposite of the previous one
        letter = self.left[signal]
        signal = self.right.find(letter)
        return signal 

File: EnigmaMachine/Java/plugboard.java
package EnigmaMachine;

public class plugboard{
    keyboard kb = new keyboard();

    String right = kb.ASCII_UPPERCASE;
    String left = kb.ASCII_UPPERCASE;


    public plugboard(String[] pairs){
        for (String i : pairs){
            char A = i.charAt(0);
            char B = i.charAt(1);

            int posA = left.indexOf(A);
            int posB = left.indexOf(B);

            left = left.substring(0, posA) + B + left.substring(posA + 1, left.length());
            left = left.substring(0, posB) + A + left.substring(posB + 1, left.length());

        }
    }

    public int forward(int signal){
        char letter = right.charAt(signal);
        signal = left.indexOf(letter);
        return signal;
    }

    public int backward(int signal){
        char letter = left.charAt(signal);
        signal = right.indexOf(letter);
        return signal;
    }
}
